server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/scm/ScmInfoRepositoryImpl.java

```diff
@@ -19,102 +19,110 @@
+    ScannerReport.Changesets changesets = scannerReportReader.readChangesets(component.getReportAttributes().getRef());
+
+    if (changesets != null) {
+      if (changesets.getChangesetCount() == 0) {
+        return generateAndMergeDb(component, changesets.getCopyFromPrevious());
+      }
+      return getScmInfoFromReport(component, changesets);
     }
-    return getScmInfoFromReport(component, changesets);
+
+    LOGGER.trace("No SCM info for file '{}'", component.getKey());
+    return generateAndMergeDb(component, false);
   }
 
-  private static ScmInfo getScmInfoFromReport(Component file, ScannerReport.Changesets changesets) {
+  private static Optional<ScmInfo> getScmInfoFromReport(Component file, ScannerReport.Changesets changesets) {
     LOGGER.trace("Reading SCM info from report for file '{}'", file.getKey());
-    return new ReportScmInfo(changesets);
+    return Optional.of(new ReportScmInfo(changesets));
   }
 
-  /**
-   * Internally used to populate cache when no ScmInfo exist.
-   */
-  enum NoScmInfo implements ScmInfo {
-    INSTANCE {
-      @Override
-      public Changeset getLatestChangeset() {
-        return notImplemented();
-      }
+  private Optional<ScmInfo> generateScmInfoForAllFile(Component file) {
+    Set<Integer> newOrChangedLines = IntStream.rangeClosed(1, file.getFileAttributes().getLines()).boxed().collect(Collectors.toSet());
+    return Optional.of(GeneratedScmInfo.create(analysisMetadata.getAnalysisDate(), newOrChangedLines));
+  }
 
-      @Override
-      public Changeset getChangesetForLine(int lineNumber) {
-        return notImplemented();
-      }
+  private static ScmInfo removeAuthorAndRevision(ScmInfo info) {
+    Map<Integer, Changeset> cleanedScmInfo = info.getAllChangesets().entrySet().stream()
+      .collect(Collectors.toMap(Map.Entry::getKey, e -> removeAuthorAndRevision(e.getValue())));
+    return new ScmInfoImpl(cleanedScmInfo);
+  }
 
-      @Override
-      public boolean hasChangesetForLine(int lineNumber) {
-        return notImplemented();
-      }
+  private static Changeset removeAuthorAndRevision(Changeset changeset) {
+    return Changeset.newChangesetBuilder().setDate(changeset.getDate()).build();
+  }
 
-      @Override
-      public Iterable<Changeset> getAllChangesets() {
-        return notImplemented();
-      }
+  private Optional<ScmInfo> generateAndMergeDb(Component file, boolean copyFromPrevious) {
+    Optional<DbScmInfo> dbInfoOpt = scmInfoDbLoader.getScmInfo(file);
+    if (!dbInfoOpt.isPresent()) {
+      return generateScmInfoForAllFile(file);
+    }
 
-      private <T> T notImplemented() {
-        throw new UnsupportedOperationException("NoScmInfo does not implement any method");
-      }
+    ScmInfo scmInfo = copyFromPrevious ? dbInfoOpt.get() : removeAuthorAndRevision(dbInfoOpt.get());
+    boolean fileUnchanged = file.getStatus() == Status.SAME && sourceHashRepository.getRawSourceHash(file).equals(dbInfoOpt.get().fileHash());
```

Comment:
why checking file hash is not enough ? In which case status can be `SAME` but file hashes are different ?

Reply:
The scanner gets the file hashes in the beginning of the analysis. The file can be changed in the DB if the CE processes a different analysis meanwhile.